File: bindings/R/FortMyrmidon/R/Space.R
#' @name fmZoneByID
#' @title a \code{std::map} of \code{\link{fmZone}} by their ID
#' @description a \code{std::map} of \code{\link{fmZone}} by their
#'     ID. It looks like an R list, but the \code{[[} R operator will
#'     only report a value if the corresponding ID is in the list, and
#'     fails otherwise.
#' @field as.list returns the map as an actual list (to iterate over)
#' @field size returns the size of the list
#' @examples
#' \dontrun{
#' antByID[[42]] # gets the fmZone with ID 42. May fail if there are no fmZone with ID 42
#' antByID.as.list # returns the actual R list, but we won't be able to query for a specific ID.
#' }
NULL

#' @name fmCZoneByID
#' @title a \code{std::map} of \code{\link{fmCZone}} by their ID
#' @description a \code{std::map} of \code{\link{fmCZone}} by their
#'     ID. It looks like an R list, but the \code{[[} R operator will
#'     only report a value if the corresponding ID is in the list, and
#'     fails otherwise.
#' @field as.list returns the map as an actual list (to iterate over)
#' @field size returns the size of the list
#' @examples
#' \dontrun{
#' antByID[[42]] # gets the fmCZone with ID 42. May fail if there are no fmCZone with ID 42
#' antByID.as.list # returns the actual R list, but we won't be able to query for a specific ID.
#' }
NULL

#' @name fmSpace
#' @title a Reference Class wrapper for
#'     \href{https://formicidae-tracker.github.io/studio/docs/latest/api/#fort::myrmidon::Space/fort::myrmidon::Space}{fort::myrmidon::Space}
#'
#' @field spaceID
#'     \href{https://formicidae-tracker.github.io/studio/docs/latest/api/#fort::myrmidon::Space/fort::myrmidon::Space::SpaceID}{fort::myrmidon::Space::SpaceID}
#' @field name
#'     \href{https://formicidae-tracker.github.io/studio/docs/latest/api/#fort::myrmidon::Space/fort::myrmidon::Space::Name}{fort::myrmidon::Space::Name}
#' @field setName
#'     \href{https://formicidae-tracker.github.io/studio/docs/latest/api/#fort::myrmidon::Space/fort::myrmidon::Space::SetName}{fort::myrmidon::Space::SetName}
#' @field createZone
#'     \href{https://formicidae-tracker.github.io/studio/docs/latest/api/#fort::myrmidon::Space/fort::myrmidon::Space::CreateZone}{fort::myrmidon::Space::CreateZone}
#' @field deleteZone
#'     \href{https://formicidae-tracker.github.io/studio/docs/latest/api/#fort::myrmidon::Space/fort::myrmidon::Space::DeleteZone}{fort::myrmidon::Space::DeleteZone}
#' @field cZones returns a \linkS4class{fmCZoneByID} as
#'     \href{https://formicidae-tracker.github.io/studio/docs/latest/api/#fort::myrmidon::Space/fort::myrmidon::Space::CZones}{fort::myrmidon::Space::CZones}
#' @field zones returns a \linkS4class{fmZoneByID} as
#'     \href{https://formicidae-tracker.github.io/studio/docs/latest/api/#fort::myrmidon::Space/fort::myrmidon::Space::Zones}{fort::myrmidon::Space::Zones}
#' @field locateMovieFrame
#'     \href{https://formicidae-tracker.github.io/studio/docs/latest/api/#fort::myrmidon::Space/fort::myrmidon::Space::LocateMovieFrame}{fort::myrmidon::Space::LocateMovieFrame}
NULL

#' @name fmCSpace
#' @title a Reference Class wrapper for
#'     \href{https://formicidae-tracker.github.io/studio/docs/latest/api/#fort::myrmidon::CSpace/fort::myrmidon::CSpace}{fort::myrmidon::CSpace}
#'
#' @field spaceID
#'     \href{https://formicidae-tracker.github.io/studio/docs/latest/api/#fort::myrmidon::CSpace/fort::myrmidon::CSpace::CSpaceID}{fort::myrmidon::CSpace::CSpaceID}
#' @field name
#'     \href{https://formicidae-tracker.github.io/studio/docs/latest/api/#fort::myrmidon::CSpace/fort::myrmidon::CSpace::Name}{fort::myrmidon::CSpace::Name}
#' @field cZones returns a \linkS4class{fmCZoneByID} as
#'     \href{https://formicidae-tracker.github.io/studio/docs/latest/api/#fort::myrmidon::CSpace/fort::myrmidon::CSpace::CZones}{fort::myrmidon::CSpace::CZones}
#' @field locateMovieFrame
#'     \href{https://formicidae-tracker.github.io/studio/docs/latest/api/#fort::myrmidon::CSpace/fort::myrmidon::CSpace::LocateMovieFrame}{fort::myrmidon::CSpace::LocateMovieFrame}
NULL

File: bindings/R/FortMyrmidon/src/space.cpp
#include "space.h"

#include "zone.h"
#include "time.h"


namespace Rcpp {
template <> SEXP wrap(const std::pair<std::string,uint64_t> & locatedMovieFrame);

template <> SEXP wrap(const std::map<fort::myrmidon::ZoneID,fort::myrmidon::CZone> & v);
template <> SEXP wrap(const std::map<fort::myrmidon::ZoneID,fort::myrmidon::Zone> & v);

}



#include <Rcpp.h>



void fmCSpace_show(const fort::myrmidon::CSpace * s) {
	Rcpp::Rcout << "fmCSpace ( ID = " << s->SpaceID()
	            << " Name = " << s->Name()
	            << " )\n";
}

void fmSpace_show(const fort::myrmidon::Space * s) {
	Rcpp::Rcout << "fmSpace ( ID = " << s->SpaceID()
	            << " Name = " << s->Name()
	            << " )\n";
}

RCPP_MODULE(space) {

	Rcpp::class_<fort::myrmidon::CSpace>("fmCSpace")
		.const_method("show",&fmCSpace_show)
		.const_method("spaceID",&fort::myrmidon::CSpace::SpaceID)
		.const_method("name",&fort::myrmidon::CSpace::Name)
		.const_method("cZones",&fort::myrmidon::CSpace::CZones)
		.const_method("locateMovieFrame",&fort::myrmidon::CSpace::LocateMovieFrame)
		;

	Rcpp::class_<fort::myrmidon::Space>("fmSpace")
		.const_method("show",&fmSpace_show)
		.const_method("spaceID",&fort::myrmidon::Space::SpaceID)
		.const_method("name",&fort::myrmidon::Space::Name)
		.method("setName",&fort::myrmidon::Space::SetName)
		.method("createZone",&fort::myrmidon::Space::CreateZone)
		.method("deleteZone",&fort::myrmidon::Space::DeleteZone)
		.const_method("cZones",&fort::myrmidon::Space::CZones)
		.method("zones",&fort::myrmidon::Space::Zones)
		.const_method("locateMovieFrame",&fort::myrmidon::Space::LocateMovieFrame)
		;

}

namespace Rcpp {

template <> SEXP wrap(const std::pair<std::string,uint64_t> & locatedMovieFrame) {
	List res;
	res["absoluteFilePath"] = locatedMovieFrame.first;
	res["movieFrameID"] = locatedMovieFrame.second;
	return res;
}

template <> SEXP wrap(const std::map<fort::myrmidon::ZoneID,fort::myrmidon::CZone> & v) {
	List res,objects(v.size());
	CharacterVector names(v.size());
	IntegerVector ID(v.size());
	int i = 0;
	for ( const auto & [zoneID,zone] : v ) {
		objects[i] = zone;
		names[i] = zone.Name();
		ID[i] = zoneID;
		++i;
	}

	res["summary"] = DataFrame::create(_["name"] = names,
	                                   _["ID"] = ID);
	res["objects"] = objects;
	return res;
}

template <> SEXP wrap(const std::map<fort::myrmidon::ZoneID,fort::myrmidon::Zone> & v) {
	List res,objects(v.size());
	CharacterVector names(v.size());
	IntegerVector ID(v.size());
	int i = 0;
	for ( const auto & [zoneID,zone] : v ) {
		objects[i] = zone;
		names[i] = zone.Name();
		ID[i] = zoneID;
		++i;
	}

	res["summary"] = DataFrame::create(_["name"] = names,
	                                   _["ID"] = ID);
	res["objects"] = objects;
	return res;
}

}
